feature: Add control plane listener

gianluca-mascolo opened the pull request:
# Description

As per [KIP-291](https://cwiki.apache.org/confluence/display/KAFKA/KIP-291%3A+Separating+controller+connections+and+requests+from+the+data+plane) Kafka Controller should have a dedicated listener.
This PR introduce new variables:
- `kafka_broker_configure_control_plane_listener` (default: `false`)
- `kafka_broker_control_plane_listener_name` (default: `controller`)

`kafka_broker_default_listeners` generation is now managed using jinja2 templating.

When `kafka_broker_configure_control_plane_listener` is set to `true` _and_ multiple listeners are enabled, the default listeners will contain a dedicated listener for the control plane. This will configure the additional listener for control plane and generate server property `control.plane.listener.name`.

Fixes #555

Altough issue [555](https://github.com/confluentinc/cp-ansible/issues/555) request this feature to be enabled by default, enabling an additional listening ports may cause issues i.e. if the playbook is used to upgrade a running cluster and the additional port is not enabled on firewalls or security groups. Leaving it disabled by default is therefore a conservative choice to make this PR a non breaking change.

In case the default port number (`8089`) assigned to control plane listener is in use by other services it can be changed customizing the `kafka_broker_custom_listeners` and adding a controller listener with the desired port, e.g.
```
kafka_broker_custom_listeners:
  controller: 
    port: 9094
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

Set `kafka_broker_configure_control_plane_listener: true` on molecule scenario `plain-rhel` and verify property `control.plane.listener.name=CONTROLLER`. Resulting `server.properties` contains:
```
control.plane.listener.name=CONTROLLER
listener.security.protocol.map=INTERNAL:SASL_PLAINTEXT,BROKER:SASL_PLAINTEXT,CONTROLLER:SASL_PLAINTEXT
listeners=INTERNAL://:9092,BROKER://:9091,CONTROLLER://:9089
```

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] Any variable changes have been validated to be backwards compatible

domenicbove commented on roles/confluent.test/molecule/plain-rhel/molecule.yml:
```diff
@@ -99,7 +99,7 @@ provisioner:
       all:
         sasl_protocol: plain
-
+        kafka_broker_configure_control_plane_listener: true
```
lets add some tasks to the verify.yml that checks the properties are there

gianluca-mascolo replied:
I've done it [here](https://github.com/confluentinc/cp-ansible/blob/fcf0bad0420a5a06d7dc11620144de7808ff6cee/roles/confluent.test/molecule/plain-rhel/verify.yml#L46). Would you like me to add some additional tests?

domenicbove commented on roles/confluent.variables/defaults/main.yml:
```diff
@@ -307,13 +307,27 @@ kafka_broker_custom_listeners:
+### Control Planer listener name.
+kafka_broker_control_plane_listener_name: "controller"
+
+kafka_broker_listeners: "{{ kafka_broker_default_listeners | combine(kafka_broker_custom_listeners | combine(kafka_broker_control_plane_listener,recursive=True) if kafka_broker_configure_control_plane_listener|bool else kafka_broker_custom_listeners, recursive=True) if kafka_broker_configure_multiple_listeners|bool else  kafka_broker_default_listeners}}"
```
I really wish there was a cleaner way to do this... at the minimum can you split it over a few lines so its easier to read. I'm going to try out something as well

```suggestion
kafka_broker_listeners: "{{ kafka_broker_default_listeners | combine(kafka_broker_custom_listeners | combine(kafka_broker_control_plane_listener,recursive=True) if kafka_broker_configure_control_plane_listener|bool 
else kafka_broker_custom_listeners, recursive=True) if kafka_broker_configure_multiple_listeners|bool 
else  kafka_broker_default_listeners}}"
```

gianluca-mascolo replied:
Yes I wish it too. I didn't know variables can go on multiple lines when in curly braces. Changed as per suggestion.

domenicbove commented:
Hey @gianluca-mascolo check this out: https://github.com/gianluca-mascolo/cp-ansible/pull/1/files

I jumped the gun haha and did some of my requests. but the big change is to use a jinja template instead of all the combines. LMK what you think

gianluca-mascolo commented:
> Hey @gianluca-mascolo check this out: https://github.com/gianluca-mascolo/cp-ansible/pull/1/files
> 
> I jumped the gun haha and did some of my requests. but the big change is to use a jinja template instead of all the combines. LMK what you think

I like it but removing `kafka_broker_custom_listeners` will result in a breaking change and we lose customization of additional listeners. Commented on your PR.

gianluca-mascolo commented:
> > Hey @gianluca-mascolo check this out: https://github.com/gianluca-mascolo/cp-ansible/pull/1/files
> > I jumped the gun haha and did some of my requests. but the big change is to use a jinja template instead of all the combines. LMK what you think
> 
> I like it but removing `kafka_broker_custom_listeners` will result in a breaking change and we lose customization of additional listeners. Commented on your PR.

It is anyI will put this PR to WIP and I will try to rework it with jinja2

> > Hey @gianluca-mascolo check this out: https://github.com/gianluca-mascolo/cp-ansible/pull/1/files
> > I jumped the gun haha and did some of my requests. but the big change is to use a jinja template instead of all the combines. LMK what you think
> 
> I like it but removing `kafka_broker_custom_listeners` will result in a breaking change and we lose customization of additional listeners. Commented on your PR.

If you prefer I can put this PR to WIP and rework it tomorrow with jinja2

gianluca-mascolo commented:
> > Hey @gianluca-mascolo check this out: https://github.com/gianluca-mascolo/cp-ansible/pull/1/files
> > I jumped the gun haha and did some of my requests. but the big change is to use a jinja template instead of all the combines. LMK what you think
> 
> I like it but removing `kafka_broker_custom_listeners` will result in a breaking change and we lose customization of additional listeners. Commented on your PR.

I will change this PR to WIP and rework it tomorrow with jinja2 templating.

gianluca-mascolo commented:
Hi @domenicbove thanks for your suggestions. I've integrated them in this PR. I found a problem in `server.properties` generation and fixed it. PR should be ready now.

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -312,6 +312,13 @@ kafka_broker_properties:
+  controlplane_listener:
+    enabled: "{{ kafka_broker_configure_control_plane_listener|bool and kafka_broker_configure_multiple_listeners|bool }}"
+    properties:
+      # The property must be defined regardless if it is enabled or not.
```
this comment doesnt match the logic in line 316. if `kafka_broker_configure_control_plane_listener = false` or `kafka_broker_configure_multiple_listeners = false`, the property `control.plane.listener.name` will not get set

gianluca-mascolo replied:
Comment is now removed because the logic was put in variable definition so that `kafka_broker_control_plane_listener_name` always point to a valid listener definition.

domenicbove commented on roles/confluent.variables/defaults/main.yml:
```diff
@@ -291,28 +291,46 @@ zookeeper_custom_properties: "{{ zookeeper.properties }}"
+kafka_broker_configure_control_plane_listener: false
+
+### Control Planer listener name.
+kafka_broker_control_plane_listener_name: controller
```
maybe there should be logic on 332, similar to 335. Thats if `control.plane.listener.name` should always get set to something

gianluca-mascolo replied:
oh yes, if I put the logic here like in [335](https://github.com/confluentinc/cp-ansible/blob/d8c40238df3df095302f486bf11c1cc74e5a9f9e/roles/confluent.variables/defaults/main.yml#L311) we can completely avoid to do the hack in the properties definition because this will always be set to something valid.

domenicbove commented:
Hey this is great! I added a question on the default behavior. Also can you double check the PR description is correct?

gianluca-mascolo commented:
> Hey this is great! I added a question on the default behavior. Also can you double check the PR description is correct?

I changed the default variable list in description. Did you mean to change something in particular?

domenicbove commented on roles/confluent.variables/defaults/main.yml:
```diff
@@ -291,28 +291,45 @@ zookeeper_custom_properties: "{{ zookeeper.properties }}"
+kafka_broker_configure_control_plane_listener: false
+
+### Control Planer listener name.
+kafka_broker_control_plane_listener_name: "{{ 'controller' if kafka_broker_configure_control_plane_listener|bool else kafka_broker_inter_broker_listener_name }}"
```
```suggestion
kafka_broker_control_plane_listener_name: "{{ 'controller' if kafka_broker_configure_control_plane_listener|bool and kafka_broker_configure_multiple_listeners|bool else kafka_broker_inter_broker_listener_name }}"
```

domenicbove replied:
we want this logic if you always want this property: `control.plane.listener.name` set. otherwise could just keep it equal to 'controller'

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -312,6 +312,10 @@ kafka_broker_properties:
                             plain_jaas_config, kafka_broker_keytab_path, kafka_broker_kerberos_principal|default('kafka'), kerberos_kafka_broker_primary,
                             sasl_scram_users.admin.principal, sasl_scram_users.admin.password, rbac_enabled_public_pem_path ) }}"
+  controlplane_listener:
+    enabled: "{{ kafka_broker_configure_control_plane_listener|bool and kafka_broker_configure_multiple_listeners|bool }}"
```
So the comment made me think this property should be enabled always, now without the comment this is fine. but just want to make sure

domenicbove replied:
I'll rephrase

Should `control.plane.listener.name` be set in the server.properties no matter what? OR set the property only when you enable it?

domenicbove replied:
Take `inter.broker.listener.name` as an example, it is under

```
  defaults:
    enabled: true
    properties:
      inter.broker.listener.name: "{{kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['name']}}"
```

Which means it will always get set. This one is only conditionally `enabled` and I'm wondering if you intended for it be always enabled? IE move it up in the same sub dictionary as `inter.broker.listener.name`

gianluca-mascolo replied:
No, It is not my intention to have control.plane.listener.name always enabled. From KIP-291 documentation

>  Whenever the "control.plane.listener.name" is set, upon broker startup, we will validate its value and make sure it's different from the inter-broker-listener-name value.
>     If the "control.plane.listener.name" config is not set, the controller will fall back to the current behavior and use inter-broker-listener-name value to determine controller-to-broker endpoints.

Therefore I think it is better to have it conditionally enabled.

gianluca-mascolo reviewed (commented):
No, It is not my intention to have `control.plane.listener.name` always enabled. From [KIP-291](https://cwiki.apache.org/confluence/display/KAFKA/KIP-291%3A+Separating+controller+connections+and+requests+from+the+data+plane) documentation

> Whenever the "control.plane.listener.name" is set, upon broker startup, we will validate its value and make sure it's different from the inter-broker-listener-name value.
> 
> If the "control.plane.listener.name" config is not set, the controller will fall back to the current behavior and use inter-broker-listener-name value to determine controller-to-broker endpoints.

Therefore I think it is better to have it conditionally enabled.

domenicbove commented on roles/confluent.variables/defaults/main.yml:
```diff
@@ -291,28 +291,45 @@ zookeeper_custom_properties: "{{ zookeeper.properties }}"
+kafka_broker_configure_control_plane_listener: false
+
+### Control Planer listener name.
+kafka_broker_control_plane_listener_name: "{{ 'controller' if kafka_broker_configure_control_plane_listener|bool and kafka_broker_configure_multiple_listeners|bool else kafka_broker_inter_broker_listener_name }}"
```
Ok this is looking good, final question, should this just be 
```suggestion
kafka_broker_control_plane_listener_name: controller
```

The if else *i think* is only necessary if you want to set the property always.

gianluca-mascolo replied:
No. Initially this was set only to 'controller' but this break the build when `kafka_broker_configure_control_plane_listener` is `false` because the playbook fails with undefined variables. (`kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name']` will be undefined in that case).

domenicbove replied:
Ahh makes sense, that is one flaw with the combine dictionary solution that we have now... wish it didn't matter... let's call it tech debt!

I wonder if this would work:
```
control.plane.listener.name: "{{kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name'] | default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name]['name'])}}"
```

or
```
control.plane.listener.name: "{{kafka_broker_listeners[kafka_broker_control_plane_listener_name]|default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name])['name']}}"
```

gianluca-mascolo replied:
You're right that should work and it is possibly more elegant that using a conditional. Let's try to change it, will take only a few mins.

domenicbove replied:
it might be 
```
control.plane.listener.name: "{{(kafka_broker_listeners[kafka_broker_control_plane_listener_name]['name'] | default(kafka_broker_listeners[kafka_broker_inter_broker_listener_name])).get('name')}}"
```

This might not be more elegant now that i try and type it out haha, wdyt?

gianluca-mascolo replied:
I'm trying the first of the two suggestions. Just running molecule to check if it works, take a while on my laptop.

domenicbove replied:
the first one is my vote (if it works)!

gianluca-mascolo replied:
yes it does! :+1:

domenicbove reviewed (approved):
Awesome, thanks for working w me on this one! Glad we got to improve the listeners implementation

gianluca-mascolo commented:
> Awesome, thanks for working w me on this one! Glad we got to improve the listeners implementation

Thanks you
